[clang-repl] Lay the foundation of pretty printing for C.

vgvassilev opened the pull request:
Depends on #89804.

vgvassilev committed:
[clang-repl] Lay the foundation of pretty printing for C.

AaronBallman reviewed (commented):
Generally looks good, but it seems that some parts of this PR are also incorporated into another PR, so it's a bit hard to review.

vgvassilev commented on clang/lib/Interpreter/Interpreter.cpp:
```diff
@@ -42,6 +42,9 @@
 #include "llvm/Support/raw_ostream.h"
 #include "llvm/TargetParser/Host.h"
+
+#include <stdarg.h>
```
Fixed.

vgvassilev commented:
> Generally looks good, but it seems that some parts of this PR are also incorporated into another PR, so it's a bit hard to review.

Rebased, should be only the commit in question now, as the other one landed.

AaronBallman commented on clang/lib/Interpreter/Interpreter.cpp:
```diff
@@ -269,7 +267,10 @@ const char *const Runtimes = R"(
     void __clang_Interpreter_SetValueCopyArr(const T (*Src)[N], void* Placement, unsigned long Size) {
       __clang_Interpreter_SetValueCopyArr(Src[0], Placement, Size);
     }
+    extern "C"
```
This seems a little subtle, but it is correct. One improvement to consider would be:
```
#ifdef __cplusplus
#define EXTERN_C extern "C"
#else
#define EXTERN_C extern
#endif
```
and using `EXTERN_C` in the declarations as needed. The current form is subtle because it's easy to miss the `extern "C"` in the prior conditional block.

vgvassilev replied:
Done.

AaronBallman reviewed (approved):
LGTM, feel free to apply my suggestion if you like it.

vgvassilev commented:
> LGTM, feel free to apply my suggestion if you like it.

Applied it. Thanks for the review this unblock whole bunch of work for us..

DavidSpickett commented:
I have bisected a 32 bit Arm failure down to this change: https://lab.llvm.org/buildbot/#/builders/245/builds/25526

Is it possible that something in this code is truncating a value when interpreting types intended for a 64 bit system, on a 32 bit system? I don't see this failure on any of our Arm64 buildbots.

vgvassilev commented:
Is the bot configured to build for arm64 and run on arm32 somehow? But in either case the width of the `int` type should be the same, right?

DavidSpickett commented:
It is a 32 bit container running on an arm64 host. As far as the compiler and tests know, it's a 32 bit system.

> But in either case the width of the int type should be the same, right?

Yes.

So I am wondering if someone of the assumptions about types being larger than one another do not hold on 32 bit. Or your change has exposed an existing issue, it wouldn't be the first time.

vgvassilev commented:
> So I am wondering if someone of the assumptions about types being larger than one another do not hold on 32 bit. Or your change has exposed an existing issue, it wouldn't be the first time.

My feeling is the latter but how can we get a debug build and debug?

DavidSpickett commented:
> My feeling is the latter but how can we get a debug build and debug?

I'm going to try a UBSAN build on the chance it pinpoints the issue right away, but failing that it'll be cross compiling to Arm 32 bit and using qemu-user to run the unit test binary. I will get you instructions on how to do that (will be tomorrow at this point).

If this is a generic 32 bit problem you could also build for 32 bit on x86. I don't know if it is because there are no bots for that.

vitalybuka commented:
It's broken here https://lab.llvm.org/buildbot/#/builders/236/builds/11633

vgvassilev commented:
> It's broken here https://lab.llvm.org/buildbot/#/builders/236/builds/11633

It seems broken for a different reason. Somehow we did not export that `__clang_Interpreter_SetValueNoAlloc` symbol. Is that some `-fvisibility=hidden` build?

vitalybuka commented:
Not sure, nothing in cmake cmd:

```
cmake -DLLVM_APPEND_VC_REV=OFF -GNinja -DCMAKE_BUILD_TYPE=Release -DLLVM_CCACHE_BUILD=ON -DLLVM_USE_LINKER=lld -DLLVM_ENABLE_ASSERTIONS=ON -DLLVM_ENABLE_PER_TARGET_RUNTIME_DIR=OFF -DCMAKE_C_COMPILER=/b/sanitizer-aarch64-linux-bootstrap-hwasan/build/llvm_build0/bin/clang -DCMAKE_CXX_COMPILER=/b/sanitizer-aarch64-linux-bootstrap-hwasan/build/llvm_build0/bin/clang++ -DLLVM_ENABLE_PLUGINS=OFF '-DLLVM_ENABLE_PROJECTS='\''clang;lld;mlir'\''' -DLLVM_USE_SANITIZER=HWAddress -DLLVM_ENABLE_LIBCXX=ON '-DCMAKE_C_FLAGS=-nostdinc++ -isystem /b/sanitizer-aarch64-linux-bootstrap-hwasan/build/libcxx_build_hwasan/include -isystem /b/sanitizer-aarch64-linux-bootstrap-hwasan/build/libcxx_build_hwasan/include/c++/v1 -fsanitize=hwaddress -mllvm -hwasan-use-after-scope=1 -Wl,--rpath=/b/sanitizer-aarch64-linux-bootstrap-hwasan/build/libcxx_build_hwasan/lib -L/b/sanitizer-aarch64-linux-bootstrap-hwasan/build/libcxx_build_hwasan/lib -w' '-DCMAKE_CXX_FLAGS=-nostdinc++ -isystem /b/sanitizer-aarch64-linux-bootstrap-hwasan/build/libcxx_build_hwasan/include -isystem /b/sanitizer-aarch64-linux-bootstrap-hwasan/build/libcxx_build_hwasan/include/c++/v1 -fsanitize=hwaddress -mllvm -hwasan-use-after-scope=1 -Wl,--rpath=/b/sanitizer-aarch64-linux-bootstrap-hwasan/build/libcxx_build_hwasan/lib -L/b/sanitizer-aarch64-linux-bootstrap-hwasan/build/libcxx_build_hwasan/lib -w' '-DCMAKE_EXE_LINKER_FLAGS=-Wl,--rpath=/b/sanitizer-aarch64-linux-bootstrap-hwasan/build/libcxx_build_hwasan/lib -L/b/sanitizer-aarch64-linux-bootstrap-hwasan/build/libcxx_build_hwasan/lib' /b/sanitizer-aarch64-linux-bootstrap-hwasan/build/llvm-project/llvm
```

vgvassilev commented:
Very strange. I did not see a lot of platforms failing. If we decide to keep that commit, is there a way to disable this test for exactly that platform?

vgvassilev commented:
> Not sure, nothing in cmake cmd:
> 
> ```
> cmake -DLLVM_APPEND_VC_REV=OFF -GNinja -DCMAKE_BUILD_TYPE=Release -DLLVM_CCACHE_BUILD=ON -DLLVM_USE_LINKER=lld -DLLVM_ENABLE_ASSERTIONS=ON -DLLVM_ENABLE_PER_TARGET_RUNTIME_DIR=OFF -DCMAKE_C_COMPILER=/b/sanitizer-aarch64-linux-bootstrap-hwasan/build/llvm_build0/bin/clang -DCMAKE_CXX_COMPILER=/b/sanitizer-aarch64-linux-bootstrap-hwasan/build/llvm_build0/bin/clang++ -DLLVM_ENABLE_PLUGINS=OFF '-DLLVM_ENABLE_PROJECTS='\''clang;lld;mlir'\''' -DLLVM_USE_SANITIZER=HWAddress -DLLVM_ENABLE_LIBCXX=ON '-DCMAKE_C_FLAGS=-nostdinc++ -isystem /b/sanitizer-aarch64-linux-bootstrap-hwasan/build/libcxx_build_hwasan/include -isystem /b/sanitizer-aarch64-linux-bootstrap-hwasan/build/libcxx_build_hwasan/include/c++/v1 -fsanitize=hwaddress -mllvm -hwasan-use-after-scope=1 -Wl,--rpath=/b/sanitizer-aarch64-linux-bootstrap-hwasan/build/libcxx_build_hwasan/lib -L/b/sanitizer-aarch64-linux-bootstrap-hwasan/build/libcxx_build_hwasan/lib -w' '-DCMAKE_CXX_FLAGS=-nostdinc++ -isystem /b/sanitizer-aarch64-linux-bootstrap-hwasan/build/libcxx_build_hwasan/include -isystem /b/sanitizer-aarch64-linux-bootstrap-hwasan/build/libcxx_build_hwasan/include/c++/v1 -fsanitize=hwaddress -mllvm -hwasan-use-after-scope=1 -Wl,--rpath=/b/sanitizer-aarch64-linux-bootstrap-hwasan/build/libcxx_build_hwasan/lib -L/b/sanitizer-aarch64-linux-bootstrap-hwasan/build/libcxx_build_hwasan/lib -w' '-DCMAKE_EXE_LINKER_FLAGS=-Wl,--rpath=/b/sanitizer-aarch64-linux-bootstrap-hwasan/build/libcxx_build_hwasan/lib -L/b/sanitizer-aarch64-linux-bootstrap-hwasan/build/libcxx_build_hwasan/lib' /b/sanitizer-aarch64-linux-bootstrap-hwasan/build/llvm-project/llvm
> ```

@vitalybuka could it be a ccache glitch?

vitalybuka commented:
> @vitalybuka could it be a ccache glitch?

Technically can, but I would not count on hash collision :)
Also there are two machines running same config consistently fail.

vitalybuka commented:
Try `// UNSUPPORTED: hwasan`

vitalybuka commented:
Actually I will try myself, as I have access to the bot.

DavidSpickett commented:
> I will get you instructions on how to do that (will be tomorrow at this point).

https://github.com/llvm/llvm-project/issues/94741#issuecomment-2154656315

There is some underlying issue (https://github.com/llvm/llvm-project/issues/94741), so if you want to try and fix it, that's great, but it's also fine if you'd rather wait and see what I find on native hardware.

vgvassilev commented:
Could be. Is there a way for me to debug this?

vgvassilev commented:
Perhaps we can disable this platform if we have a way to express the setup in lit.

DavidSpickett commented:
You can debug it by cross-compiling a debug build, then running the unit tests binary and connecting to qemu's built in gdbserver. The Arm toolchain might have a copy of gdb in it already, but if not, gdb multiarch and lldb would also work (which are both installable from apt).

I'm going to try looking in a debugger myself today.

vgvassilev commented:
> You can debug it by cross-compiling a debug build, then running the unit tests binary and connecting to qemu's built in gdbserver. The Arm toolchain might have a copy of gdb in it already, but if not, gdb multiarch and lldb would also work (which are both installable from apt).
> 
> I'm going to try looking in a debugger myself today.

Ok, this sounds good. I've never thought about that setup and perhaps is just fine to suppress that test.

zmodem commented:
We're seeing the test fail on our mac builds: https://issues.chromium.org/issues/346289767

```
JIT session error: Symbols not found: [ ___clang_Interpreter_SetValueNoAlloc ]
 error: Failed to materialize symbols: { (main, { ____orc_init_func.incr_module_10, _c_str, $.incr_module_10.__inits.0 }) }
 error: Failed to materialize symbols: { (main, { ____orc_init_func.incr_module_10 }) }
```

Our current theory is that it's because we configure with `-DCLANG_PLUGIN_SUPPORT=OFF` and clang-repl has this in its cmakelists.txt:

https://github.com/llvm/llvm-project/blob/f26bc5f0c421c7a01a1a611249dd7483b749bab5/clang/tools/clang-repl/CMakeLists.txt#L64-L67

Maybe it should call `export_executable_symbols_for_plugins` unconditionally? Or the new test should `REQUIRES: plugins`?

zmodem commented:
> Would dropping this conditional make it work for your case?

I don't have a mac handy to try it right now. If you commit it, our system will test it automatically :) Or maybe @thakis could give it a try?

vgvassilev commented:
> > Would dropping this conditional make it work for your case?
> 
> I don't have a mac handy to try it right now. If you commit it, our system will test it automatically :) Or maybe @Thakis could give it a try?

I am on my phone in the next couple of hours and I am not sure how to reproduce it. Maybe somebody who can should give it a shot?

zmodem commented:
I was able to confirm on a local Mac that removing the `if(CLANG_PLUGIN_SUPPORT)` fixes the test, so I'll go ahead and commit that.

aeubanks commented:
So actually even the `export_executable_symbols_for_plugins` doesn't fix our bots. I've narrowed it down to `-DLLVM_ENABLE_PIC=ON/OFF`. Perhaps we're not exporting symbols when `-DLLVM_ENABLE_PIC=OFF`.

vgvassilev commented:
> So actually even the `export_executable_symbols_for_plugins` doesn't fix our bots. I've narrowed it down to `-DLLVM_ENABLE_PIC=ON/OFF`. Perhaps we're not exporting symbols when `-DLLVM_ENABLE_PIC=OFF`.

That’s quite strange. Is that somehow related to ld64? @lhames donyiu have any clue?

aeubanks commented:
This was with `-DLLVM_ENABLE_LLD=ON`, so probably not? I'm guessing there's some CMake logic that doesn't fit with clang-repl's expectation of symbol availability, but that may be wrong.

vgvassilev commented:
Our current theory is that somehow the build system does not mark this symbol as exported. Maybe we can confirm that by making the symbol resolution for that particular one via the JIT absolute symbol definition. This is rather a workaround and the real fix should be in the build system.

```diff
diff --git a/clang/lib/Interpreter/Interpreter.cpp b/clang/lib/Interpreter/Interpreter.cpp
index 7a9527891427..4131d5ff0d99 100644
--- a/clang/lib/Interpreter/Interpreter.cpp
+++ b/clang/lib/Interpreter/Interpreter.cpp
@@ -43,6 +43,9 @@
 #include "llvm/Support/raw_ostream.h"
 #include "llvm/TargetParser/Host.h"
 
+#include "llvm/ExecutionEngine/Orc/LLJIT.h"
+
+
 #include <cstdarg>
 
 using namespace clang;
@@ -405,6 +408,83 @@ createJITTargetMachineBuilder(const std::string &TT) {
   return llvm::orc::JITTargetMachineBuilder(llvm::Triple(TT));
 }
 
+extern "C" void REPL_EXTERNAL_VISIBILITY __clang_Interpreter_SetValueNoAlloc(
+    void *This, void *OutVal, void *OpaqueType, ...) {
+  Value &VRef = *(Value *)OutVal;
+  Interpreter *I = static_cast<Interpreter *>(This);
+  VRef = Value(I, OpaqueType);
+  if (VRef.isVoid())
+    return;
+
+  va_list args;
+  va_start(args, /*last named param*/ OpaqueType);
+
+  QualType QT = VRef.getType();
+  if (VRef.getKind() == Value::K_PtrOrObj) {
+    VRef.setPtr(va_arg(args, void *));
+  } else {
+    if (const auto *ET = QT->getAs<EnumType>())
+      QT = ET->getDecl()->getIntegerType();
+    switch (QT->castAs<BuiltinType>()->getKind()) {
+    default:
+      llvm_unreachable("unknown type kind!");
+      break;
+      // Types shorter than int are resolved as int, else va_arg has UB.
+    case BuiltinType::Bool:
+      VRef.setBool(va_arg(args, int));
+      break;
+    case BuiltinType::Char_S:
+      VRef.setChar_S(va_arg(args, int));
+      break;
+    case BuiltinType::SChar:
+      VRef.setSChar(va_arg(args, int));
+      break;
+    case BuiltinType::Char_U:
+      VRef.setChar_U(va_arg(args, unsigned));
+      break;
+    case BuiltinType::UChar:
+      VRef.setUChar(va_arg(args, unsigned));
+      break;
+    case BuiltinType::Short:
+      VRef.setShort(va_arg(args, int));
+      break;
+    case BuiltinType::UShort:
+      VRef.setUShort(va_arg(args, unsigned));
+      break;
+    case BuiltinType::Int:
+      VRef.setInt(va_arg(args, int));
+      break;
+    case BuiltinType::UInt:
+      VRef.setUInt(va_arg(args, unsigned));
+      break;
+    case BuiltinType::Long:
+      VRef.setLong(va_arg(args, long));
+      break;
+    case BuiltinType::ULong:
+      VRef.setULong(va_arg(args, unsigned long));
+      break;
+    case BuiltinType::LongLong:
+      VRef.setLongLong(va_arg(args, long long));
+      break;
+    case BuiltinType::ULongLong:
+      VRef.setULongLong(va_arg(args, unsigned long long));
+      break;
+      // Types shorter than double are resolved as double, else va_arg has UB.
+    case BuiltinType::Float:
+      VRef.setFloat(va_arg(args, double));
+      break;
+    case BuiltinType::Double:
+      VRef.setDouble(va_arg(args, double));
+      break;
+    case BuiltinType::LongDouble:
+      VRef.setLongDouble(va_arg(args, long double));
+      break;
+      // See REPL_BUILTIN_TYPES.
+    }
+  }
+  va_end(args);
+}
+
 llvm::Error Interpreter::CreateExecutor() {
   if (IncrExecutor)
     return llvm::make_error<llvm::StringError>("Operation failed. "
@@ -431,6 +511,12 @@ llvm::Error Interpreter::CreateExecutor() {
   if (!Err)
     IncrExecutor = std::move(Executor);
 
+  auto &J = IncrExecutor->GetExecutionEngine();
+  Err = J.getMainJITDylib().define(
+                                   llvm::orc::absoluteSymbols({{J.mangleAndIntern("__clang_Interpreter_SetValueNoAlloc"),
+                                                    {llvm::orc::ExecutorAddr::fromPtr(&__clang_Interpreter_SetValueNoAlloc),
+                                                     llvm::JITSymbolFlags::Exported}}}));
+
   return Err;
 }
 
@@ -860,83 +946,6 @@ __clang_Interpreter_SetValueWithAlloc(void *This, void *OutVal,
   return VRef.getPtr();
 }
 
-extern "C" void REPL_EXTERNAL_VISIBILITY __clang_Interpreter_SetValueNoAlloc(
-    void *This, void *OutVal, void *OpaqueType, ...) {
-  Value &VRef = *(Value *)OutVal;
-  Interpreter *I = static_cast<Interpreter *>(This);
-  VRef = Value(I, OpaqueType);
-  if (VRef.isVoid())
-    return;
-
-  va_list args;
-  va_start(args, /*last named param*/ OpaqueType);
-
-  QualType QT = VRef.getType();
-  if (VRef.getKind() == Value::K_PtrOrObj) {
-    VRef.setPtr(va_arg(args, void *));
-  } else {
-    if (const auto *ET = QT->getAs<EnumType>())
-      QT = ET->getDecl()->getIntegerType();
-    switch (QT->castAs<BuiltinType>()->getKind()) {
-    default:
-      llvm_unreachable("unknown type kind!");
-      break;
-      // Types shorter than int are resolved as int, else va_arg has UB.
-    case BuiltinType::Bool:
-      VRef.setBool(va_arg(args, int));
-      break;
-    case BuiltinType::Char_S:
-      VRef.setChar_S(va_arg(args, int));
-      break;
-    case BuiltinType::SChar:
-      VRef.setSChar(va_arg(args, int));
-      break;
-    case BuiltinType::Char_U:
-      VRef.setChar_U(va_arg(args, unsigned));
-      break;
-    case BuiltinType::UChar:
-      VRef.setUChar(va_arg(args, unsigned));
-      break;
-    case BuiltinType::Short:
-      VRef.setShort(va_arg(args, int));
-      break;
-    case BuiltinType::UShort:
-      VRef.setUShort(va_arg(args, unsigned));
-      break;
-    case BuiltinType::Int:
-      VRef.setInt(va_arg(args, int));
-      break;
-    case BuiltinType::UInt:
-      VRef.setUInt(va_arg(args, unsigned));
-      break;
-    case BuiltinType::Long:
-      VRef.setLong(va_arg(args, long));
-      break;
-    case BuiltinType::ULong:
-      VRef.setULong(va_arg(args, unsigned long));
-      break;
-    case BuiltinType::LongLong:
-      VRef.setLongLong(va_arg(args, long long));
-      break;
-    case BuiltinType::ULongLong:
-      VRef.setULongLong(va_arg(args, unsigned long long));
-      break;
-      // Types shorter than double are resolved as double, else va_arg has UB.
-    case BuiltinType::Float:
-      VRef.setFloat(va_arg(args, double));
-      break;
-    case BuiltinType::Double:
-      VRef.setDouble(va_arg(args, double));
-      break;
-    case BuiltinType::LongDouble:
-      VRef.setLongDouble(va_arg(args, long double));
-      break;
-      // See REPL_BUILTIN_TYPES.
-    }
-  }
-  va_end(args);
-}
-
 // A trampoline to work around the fact that operator placement new cannot
 // really be forward declared due to libc++ and libstdc++ declaration mismatch.
 // FIXME: __clang_Interpreter_NewTag is ODR violation because we get the same
 ```

Courtesy of @lhames.

anutosh491 commented on clang/lib/Parse/ParseStmt.cpp:
```diff
@@ -571,11 +571,8 @@ StmtResult Parser::ParseExprStatement(ParsedStmtContext StmtCtx) {
-  // since part of the implementation requires C++ language features.
   // Note we shouldn't eat the token since the callback needs it.
-  if (Tok.is(tok::annot_repl_input_end) && Actions.getLangOpts().CPlusPlus)
+  if (Tok.is(tok::annot_repl_input_end))
```
Hi,

Landed here as was studying a bit on `annot_repl_input_end`

Curious if this change also calls for a change in the docs ? @vgvassilev 

I see the following 

![image](https://github.com/user-attachments/assets/d6e22a0a-9c17-4859-a1cf-ea09124b6226)

vgvassilev replied:
Probably.

anutosh491 replied:
Cool going ahead and raising a PR for the same !
